### ISSUE TITLE
profiler: WithAgentAddr doesn't appropriately override UDS client from DD_TRACE_AGENT_URL

### ISSUE THREAD

**Issue by nsrip-dd** [DataDog/dd-trace-go#1554]
The following test fails with a timeout:

```
func TestPrecedence(t *testing.T) {
	t.Setenv("DD_TRACE_AGENT_URL", "unix:///made-up-path-that-does-not-exist")
	received := make(chan struct{}, 1)
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		select {
		case received <- struct{}{}:
		default:
		}
	}))
	defer server.Close()

	err := Start(
		WithAgentAddr(server.Listener.Addr().String()),
		WithPeriod(1*time.Second),
	)
	require.NoError(t, err)
	defer Stop()

	timeout := time.After(5 * time.Second)
	select {
	case <-timeout:
		t.Fatal("should have received a profile already")
	case <-received:
	}
}
```

If we start the profiler with the `WithAgentAddr` option, that in-app config should take precedence over the environment variable. However, providing a Unix domain socket path to `DD_TRACE_AGENT_URL` causes the profiler to use an `http.Client` with a UDS transport in [`defaultConfig`](https://github.com/DataDog/dd-trace-go/blob/6b6a35446eee2a06e6d1d1986d54f9758a7c1ee2/profiler/options.go#L228). This doesn't get overridden if there's a subsequent call to `WithAgentAddr`. I think it should.

This also raises  the question of what should happen if a user combines `WithAgentAddr` and `WithUDS`?

This might affect the tracer as well?

**Comment by mackjmr** [comment]
> This might affect the tracer as well?

Yes, there is the same behaviour on the tracer. If `DD_TRACE_AGENT_URL` is set to a valid UDS path, then we will use a UDS transport even if `WithAgentAddr` is set -- calling `WithAgentAddr` won't override the client.

> This also raises the question of what should happen if a user combines `WithAgentAddr` and `WithUDS`?

In the tracer, this has the same behaviour as setting `DD_TRACE_AGENT_URL` and `WithAgentAddr`: if they are both set, then we use a UDS transport. 

Perhaps the fix could address both of these: a call to `WithAgentAddr` should override the client if it's currently using a UDS transport. That said, this would mean the order of the in-code option matters, as this would only override if `WithAgentAddr` is set **after** `WithUDS`. In the case where `WithUDS` is set after `WithAgentAddr`, then the UDS transport will take precedence. Not sure if this is an acceptable solution ?

**Comment by nsrip-dd** [comment]
> Perhaps the fix could address both of these: a call to WithAgentAddr should override the client if it's currently using a UDS transport. That said, this would mean the order of the in-code option matters, as this would only override if WithAgentAddr is set after WithUDS. In the case where WithUDS is set after WithAgentAddr, then the UDS transport will take precedence. Not sure if this is an acceptable solution ?

I considered that as well. I think this is an unfortunate downside of using functional options. I think it would be okay if we have WithAgentAddr override the client if it was previously UDS. It's mainly important that we're consistent about whether in-code config takes precedence over environment variables. If a user supplies both of those options in the app, in either order, I'd consider that to be user error and we don't need to specifically define what happens. What do you think?

**Comment by Primetalk** [comment]
At the start of tracer an invalid message is printed regardless of UDS being used:
https://github.com/DataDog/dd-trace-go/blob/main/ddtrace/tracer/log.go#L116

**Comment by ktlacaelel** [comment]
Unix Domain Sockets do not work!
This feature is missing from the go trace agent implementation.

Reason: The logic inside WithUDS is calling WithHTTPClient so it always tries to connect via http.
Regardless of how it was configured.

When called WithUDS the unix domain socket address is wrapped around http and it obviously fails to perform an http request on a unix domain socket.

<img width="1064" alt="image" src="https://user-images.githubusercontent.com/30967/206014602-57e6f1c8-5723-465e-ac5d-ee810485e200.png">

**Comment by mackjmr** [comment]
@ktlacaelel when `WithUDS` is called we override the HTTP client to connect to the socket in its [dialer](https://github.com/DataDog/dd-trace-go/blob/274246638d1507d3f9bdc9503e5650d90612dcdf/ddtrace/tracer/option.go#L343-L348).

Can you clarify how you are setting the agent URL ?

Note: in order to use `DD_TRACE_AGENT_URL`, you would need to update to at least v1.44.0 (it seems like you are on v1.43.1).

**Comment by ktlacaelel** [comment]
Thanks for the prompt response @mackjmr here is a snippet on how we are trying to set it up!
```
 	tracer.Start(
 		tracer.WithEnv("development"),
 		tracer.WithService("dejavu"),
 		tracer.WithServiceVersion("abc123"),
        tracer.WithUDS("unix:///var/run/datadog/apm.socket"),
 	)
```

We've also set the `DD_TRACE_AGENT_URL` and have confirmed that the socket path exists and works!
But when we run the code even with the latest version: `1.44.1` we see that is still trying to connect using tcp.

<img width="1217" alt="image" src="https://user-images.githubusercontent.com/30967/206547705-a822592b-dba3-4251-b0a5-577abd3990e4.png">

**Comment by mackjmr** [comment]
@ktlacaelel thanks for sharing the additional information. The diagnostics and tracer configuration logs showing TCP is due to the issue mentioned by @Primetalk above (which is being addressed in this [PR](https://github.com/DataDog/dd-trace-go/pull/1604)). Please note that even though the startup logs show a TCP connection, the tracer will use UDS when sending traces to the trace-agent. 
Additionally, you don't need to set _both_ `WithUDS` and `DD_TRACE_AGENT_URL` -- only one of the two will do.

**Comment by ktlacaelel** [comment]
Thanks @mackjmr we've tried with either `WithUDS` and `DD_TRACE_AGENT_URL` without any luck.
In the beginning we had it configured with `DD_TRACE_AGENT_URL` only but noticed we weren't getting any metrics.
We know our agent is configured properly and that the socket is working as this is used by another app running in node that is properly reporting.
we seem to only be missing the metrics in the case of the go-tracer, once we opened the logs we noticed it was using a different protocol, any advice on how we can debug this?

**Comment by ktlacaelel** [comment]
@mackjmr thanks, in the screenshot above, we can clearly see that besides logging the tcp connection it is trying to make a connection to tcp and failling is there anything that we might be missing?
